frame/bridge/ethereum-bsc/src/lib.rs

```diff
@@ -0,0 +1,559 @@
+			ensure!(
+				!Self::is_timestamp_ahead(header.timestamp),
+				<Error<T>>::HeaderTimestampIsAhead
+			);
+
+			// Check that the extra-data contains the vanity, validators and signature.
+			ensure!(
+				header.extra_data.len() > VANITY_LENGTH,
+				<Error<T>>::MissingVanity
+			);
+
+			let validator_bytes_len = header
+				.extra_data
+				.len()
+				.checked_sub(VANITY_LENGTH + SIGNATURE_LENGTH)
+				.ok_or(<Error<T>>::MissingSignature)?;
+			// Ensure that the extra-data contains a validator list on checkpoint, but none otherwise
+			let is_checkpoint = header.number % config.epoch_length == 0;
+
+			if is_checkpoint {
+				// Checkpoint blocks must at least contain one validator
+				ensure!(
+					validator_bytes_len != 0,
+					<Error<T>>::InvalidCheckpointValidators
+				);
+				// Ensure that the validator bytes length is valid
+				ensure!(
+					validator_bytes_len % ADDRESS_LENGTH == 0,
+					<Error<T>>::InvalidCheckpointValidators
+				);
+			} else {
+				ensure!(validator_bytes_len == 0, <Error<T>>::ExtraValidators);
+			}
+
+			Ok(())
+		}
+
+		/// Perform checks that require access to parent header.
+		pub fn contextual_checks(
+			config: &BSCConfiguration,
+			header: &BSCHeader,
+			parent: &BSCHeader,
+		) -> DispatchResult {
+			// parent sanity check
+			if parent.compute_hash() != header.parent_hash || parent.number + 1 != header.number {
+				Err(<Error<T>>::UnknownAncestor)?;
+			}
+
+			// Ensure that the block's timestamp isn't too close to it's parent
+			// And header.timestamp is greater than parents'
+			if header.timestamp < parent.timestamp.saturating_add(config.period) {
+				Err(<Error<T>>::HeaderTimestampTooClose)?;
+			}
+
+			Ok(())
+		}
+
+		/// Recover block creator from signature
+		pub fn recover_creator(
+			chain_id: u64,
+			header: &BSCHeader,
+		) -> Result<Address, DispatchError> {
+			if let Some(creator) = <CreatorCache<T>>::get(header.compute_hash()) {
+				return Ok(creator);
+			}
+
+			let data = &header.extra_data;
+
+			ensure!(data.len() > VANITY_LENGTH, <Error<T>>::MissingVanity);
+			ensure!(
+				data.len() >= VANITY_LENGTH + SIGNATURE_LENGTH,
+				<Error<T>>::MissingSignature
+			);
+
+			// Split `signed_extra data` and `signature`
+			let (signed_data_slice, signature_slice) = data.split_at(data.len() - SIGNATURE_LENGTH);
+			// convert `&[u8]` to `[u8; 65]`
+			let signature = {
+				let mut s = [0; SIGNATURE_LENGTH];
+				s.copy_from_slice(signature_slice);
+
+				s
+			};
+			// modify header and hash it
+			let mut unsigned_header = header.to_owned();
+
+			unsigned_header.extra_data = signed_data_slice.to_vec();
+
+			let msg = unsigned_header.compute_hash_with_chain_id(chain_id);
+			let pubkey = crypto::secp256k1_ecdsa_recover(&signature, msg.as_fixed_bytes())
+				.map_err(|_| <Error<T>>::RecoverPubkeyFail)?;
+			let creator = bsc_primitives::public_to_address(&pubkey);
+
+			<CreatorCache<T>>::insert(header.compute_hash(), creator);
```

Comment:
This does not make sense.

`CreatorCache` is crypto recover local cache, not global, should not be stored on chain in storage.